File: tst/clip_raw.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);
    pico_set_view("Clip", -1, -1, NULL, NULL, NULL, NULL, NULL);

    puts("rect at 30%");
    pico_output_clear();
    Pico_Rel_Rect r1 = { '!', {30-25, 30-25, 50, 50}, PICO_ANCHOR_NW, NULL };
    pico_output_draw_rect(&r1);
    _pico_check("clip-01"); // same as pct

    puts("gray background");
    pico_set_view(NULL, -1, -1, NULL, NULL, NULL, NULL, &r1);
    pico_set_color_clear((Pico_Color){0xCC, 0xCC, 0xCC});
    pico_output_clear();
    _pico_check("clip-02"); // same as pct

    puts("red centered under gray");
    Pico_Rel_Rect r2 = { '!', {r1.x+r1.w/2-12, r1.y+r1.h/2-12, 25, 25}, PICO_ANCHOR_NW, NULL };
    pico_set_color_draw((Pico_Color){0xFF,0x00,0x00});
    pico_output_draw_rect(&r2);
    _pico_check("clip-03"); // same as pct

    puts("blue centered under gray");
    Pico_Rel_Pos p3 = { '!', {r1.x+r1.w/2, r1.y+r1.h/2}, PICO_ANCHOR_NW, NULL };
    pico_set_color_draw((Pico_Color){0x00,0x00,0xFF});
    pico_output_draw_pixel(&p3);
    _pico_check("clip-04"); // same as pct

    puts("yellow clipped under gray bottom right");
    Pico_Rel_Rect r4 = { '!', {r1.x+r1.w-12, r1.y+r1.h-12, 25, 25}, PICO_ANCHOR_NW, NULL };
    pico_set_color_draw((Pico_Color){0xFF,0xFF,0x00});
    pico_output_draw_rect(&r4);
    _pico_check("clip-05"); // same as pct

    pico_init(0);
    return 0;
}

File: lua/tst/clip_raw.lua
require 'pico.check'

pico.init(true)
pico.set.window { title="Clip" }

do
    print("rect at 30%")
    pico.output.clear()
    local r1 = {'!', x=30-25, y=30-25, w=50, h=50, anc='NW'}
    pico.output.draw.rect(r1)
    pico.check("clip-01")
end

local r1 = {'!', x=30-25, y=30-25, w=50, h=50, anc='NW'}

do
    print("gray background")
    pico.set.view { clip=r1 }
    pico.set.color.clear({'!', r=0xCC, g=0xCC, b=0xCC})
    pico.output.clear()
    pico.check("clip-02")
end

do
    print("red centered under gray")
    local r2 = {'!', x=r1.x+r1.w/2-12, y=r1.y+r1.h/2-12, w=25, h=25, anc='NW'}
    pico.set.color.draw('red')
    pico.output.draw.rect(r2)
    pico.check("clip-03")
end

do
    print("blue centered under gray")
    local p3 = {'!', x=r1.x+r1.w/2, y=r1.y+r1.h/2, anc='NW'}
    pico.set.color.draw('blue')
    pico.output.draw.pixel(p3)
    pico.check("clip-04")
end

do
    print("yellow clipped under gray bottom right")
    local r4 = {'!', x=r1.x+r1.w-12, y=r1.y+r1.h-12, w=25, h=25, anc='NW'}
    pico.set.color.draw('yellow')
    pico.output.draw.rect(r4)
    pico.check("clip-05")
end

pico.init(false)
